include/pci.h

```diff
@@ -81,32 +84,28 @@ struct pci_device {
 
   unsigned nbars;
   resource_t bar[6];
-};
+} pci_dev_data_t;
 
-/* TODO: pci_bus_device will become a state (device_t) of generic PCI driver. */
-typedef struct pci_bus_device {
-  pci_bus_t *bus;
-  resource_t *mem_space;
-  resource_t *io_space;
-  pci_device_list_t devices;
-} pci_bus_device_t;
+#define PCI_DRIVER(dev) ((pci_bus_driver_t *)((dev)->parent->driver))
 
-static inline uint32_t pci_read_config(pci_device_t *device, unsigned reg,
+static inline uint32_t pci_read_config(device_t *device, unsigned reg,
                                        unsigned size) {
-  return device->bus->read_config(device, reg, size);
+  return PCI_DRIVER(device)->pci_bus.read_config(device, reg, size);
 }
 
-static inline void pci_write_config(pci_device_t *device, unsigned reg,
+static inline void pci_write_config(device_t *device, unsigned reg,
                                     unsigned size, uint32_t value) {
-  device->bus->write_config(device, reg, size, value);
+  PCI_DRIVER(device)->pci_bus.write_config(device, reg, size, value);
 }
 
-static inline uint32_t pci_adjust_config(pci_device_t *device, unsigned reg,
+static inline uint32_t pci_adjust_config(device_t *device, unsigned reg,
                                          unsigned size, uint32_t value) {
   pci_write_config(device, reg, size, value);
   return pci_read_config(device, reg, size);
 }
 
-void pci_init();
+void pci_bus_enumerate(device_t *pcib);
+void pci_bus_assign_space(device_t *pcib);
+void pci_bus_dump(device_t *pcib);
```

Comment:
Okay, so what I actually meant was this interface would ideally look like this:

```c
uint32_t pci_read_config(pci_device_t *device, unsigned reg, unsigned size);
void pci_write_config(pci_device_t *device, unsigned reg, unsigned size, uint32_t value);
uint32_t pci_adjust_config(pci_device_t *device, unsigned reg, unsigned size, uint32_t value);
void pci_bus_enumerate(pci_bus_device_t *pcib);
void pci_bus_assign_space(pci_bus_device_t *pcib);
void pci_bus_dump(pci_bus_device_t *pcib);
```

In my suggested layout I would consider both `pci_device_t` and `pci_bus_device_t` extensions of `device_t`.